### PR TITLE
[espnet3-recipe] Add LibriTTS recipe for codec models

### PR DESCRIPTION
**Pull request by NewGamezzz**
## What did you change?

Add a LibriTTS codec recipe to ESPnet3, migrated from the espnet2 recipe (egs2/libritts/codec1).

---

## Why did you make this change?

This PR migrates the espnet2 LibriTTS codec recipe to ESPnet3 so codec models (e.g., Encodec) can be trained with the ESPnet3 workflow.

---

## Is your PR small enough?

This PR contains 26 file changes. However, some of them are just config files.

---

## Additional Context

This PR was tested by training an Encodec model on 4x V100 GPUs. Results on LibriTTS test-clean:
| Metrics      | Scores |
| ----------- | ----------- |
|  utmos        |             3.9874|
|  dns_overall     |          3.1150|
 | dns_p808      |            3.7710|

I modified espnet3/components/data/dataloader.py to fix a multi-GPU deadlock caused by an unequal number of batches per rank. With ChunkIterFactory, ranks are not guaranteed to yield the same number of chunks: a rank that receives longer audio splits it into more chunks. When one rank runs out of batches while another still has some, the ranks disagree on when to enter validation and training deadlocks. To fix this, I added a DistributedSyncIterator that performs an all-reduce each step to check whether any rank has exhausted its samples; if so, all ranks stop the epoch together and enter validation in sync. I also removed the per-rank batch size check, since ESPnet3 uses the same batch size on every rank and the check is no longer needed.

To be discussed: I am thinking of moving the gan module to the base directory because other recipes, such as VITS in the TTS task, can use it. We can also move versa metric to the espnet3 directory too.

### CODE REVIEW

File: espnet3/components/data/dataloader.py [hunk 1]
```diff
@@ -39,6 +39,48 @@ def log_dataloader(logger: logging.Logger, loader, label: str) -> None:
...
+            except StopIteration:
+                batch = None
+                has_next = torch.zeros(1, device=device)
+            torch.distributed.all_reduce(has_next, op=torch.distributed.ReduceOp.MIN)
+            # After the all-reduce, has_next is 0 on all ranks if any rank's
+            # iterator is exhausted.
+            if has_next.item() == 0:
+                return
+            yield batch
+
```
Comment (Emrys365) [inline espnet3/components/data/dataloader.py]: This issue was solved in ESPnet2 in a similar way, by all-reducing an `iterator_stop` flag across ranks.
> https://github.com/espnet/espnet/blob/master/espnet2/train/trainer.py#L567-L581

So I think this fix is totally fine.

File: egs3/libritts/codec/conf/metrics.yaml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+# with versa (https://github.com/wavlab-speech/versa), which must be
+# installed in the evaluation environment.
+
+dataset:
+  test:
+    - name: test
+
+metrics:
+  - metric:
+      _target_: src.metrics.versa.VersaMetric
```
Comment (Masao-Someki) [inline egs3/libritts/codec/conf/metrics.yaml]: Let's move to `espnet3/components/metrics/versa.py` and refer to that module

File: egs3/libritts/codec/conf/training_encodec.yaml [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+#   python run.py --stages create_dataset --training_config conf/training_encodec.yaml
+#   python run.py --stages train           --training_config conf/training_encodec.yaml
+#
+##########################################################
+#                  SYSTEM AND PATH SETTING               #
+##########################################################
+task: espnet2.tasks.gan_codec.GANCodecTask
+
+num_device: 1
+num_nodes: 1
```
Comment (Masao-Someki) [inline egs3/libritts/codec/conf/training_encodec.yaml]: You can remove these configs when config values are the default values. (default config values = config values under `TEMPLATE/codec/conf/`)
Let's keep recipe configs smaller and simple!

File: egs3/libritts/codec/dataset/builder.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            recipe_dir: Recipe root directory.
+            **_kwargs: Unused extra options for API compatibility.
+
+        Notes:
+            This method:
+            1. Checks for the presence of required LibriTTS subsets.
+            2. If not present, it runs the download script for each missing subset.
+            3. Verifies that the required subsets are present after the download
+               attempt.
+
```
Comment (Masao-Someki) [inline egs3/libritts/codec/dataset/builder.py:207]: Let's add example!

Comment (Masao-Someki) [reply egs3/libritts/codec/dataset/builder.py:207]: Also, you can check `https://github.com/espnet/espnet/blob/master/espnet3/utils/download_utils.py`, which provides file download support with progress logging.

File: egs3/libritts/codec/dataset/builder.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    ) -> None:
+        """Build LibriTTS dataset artifacts (manifests).
+
+        Args:
+            recipe_dir: Recipe root directory.
+            **_kwargs: Optional keyword arguments for build customization:
+
+        Returns:
+            None.
+
```
Comment (Masao-Someki) [inline egs3/libritts/codec/dataset/builder.py:287]: Let's add example!

File: egs3/libritts/codec/local/download_libritts.sh
Comment (Masao-Someki) [inline egs3/libritts/codec/local/download_libritts.sh:1]: Let's make everything work by python!

File: espnet3/components/metrics/versa.py
Comment (Masao-Someki) [inline espnet3/components/metrics/versa.py:1]: Let's move to `espnet3/components/metrics/versa.py` and refer to that module. I think this is quite useful to many other generation-related systems.

File: egs3/TEMPLATE/codec/conf/training.yaml [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+    params: discriminator        # substring used to select parameters
+    accum_grad_steps: 1          # number of backward passes before optimizer.step()
+    step_every_n_iters: 1        # only step on every N-th training iteration
+    gradient_clip_val: 1.0
+
+schedulers:
+  generator:
+    scheduler:
+      _target_: torch.optim.lr_scheduler.ExponentialLR
+      gamma: 0.999875             # scheduler-specific argument
```
Comment (Masao-Someki) [inline egs3/TEMPLATE/codec/conf/training.yaml]: Is there any specific reason to use this value for gamma?

File: egs3/TEMPLATE/codec/conf/training.yaml [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+    interval: epoch               # `epoch` schedulers run in `on_train_epoch_end()`
+
+  discriminator:
+    scheduler:
+      _target_: torch.optim.lr_scheduler.ExponentialLR
+      gamma: 0.999875
```
Comment (Masao-Someki) [inline egs3/TEMPLATE/codec/conf/training.yaml]: ditto

File: egs3/TEMPLATE/codec/conf/training.yaml [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+dataloader:
+  collate_fn:
+    _target_: espnet2.train.collate_fn.CommonCollateFn
+    float_pad_value: 0.0
+    int_pad_value: 0
+  train:
+    iter_factory:
+      _target_: espnet2.iterators.chunk_iter_factory.ChunkIterFactory
+      batch_size: 4
+      chunk_length: 61440
```
Comment (Masao-Someki) [inline egs3/TEMPLATE/codec/conf/training.yaml]: ditto, is there any reason on this number?

File: egs3/TEMPLATE/codec/conf/training.yaml [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+      batches:
+        type: unsorted
+        batch_size: 1        # required by ChunkIterFactory's per-sample sampler
+        batch_bins: 0        # required by signature, unused for "unsorted"
+        shape_files:
+          - ${data_dir}/manifest/train.tsv
+  valid:
+    iter_factory:
+      _target_: espnet2.iterators.chunk_iter_factory.ChunkIterFactory
+      batch_size: 4
```
Comment (Masao-Someki) [inline egs3/TEMPLATE/codec/conf/training.yaml]: Why not refer to the training config? I thought we could simply use the values defined there.

File: egs3/TEMPLATE/codec/run.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+    run_stages,
+)
+
+# Default stage list (can be extended/overridden by callers)
+DEFAULT_STAGES: List[str] = [
+    "create_dataset",
+    "collect_stats",
+    "train",
+    "infer",
+    "measure",
```
Comment (Masao-Someki) [inline egs3/TEMPLATE/codec/run.py:30]: let's add publication-related, demo-related stages~

File: espnet3/systems/codec/gan_trainer.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+        super().__init__(
+            model=model,
+            exp_dir=exp_dir,
+            config=trainer_config,
+            best_model_criterion=best_model_criterion,
+        )
+
+
+def build_gan_trainer(training_config, model) -> GANLightningTrainer:
+    """Build the shared GAN Lightning trainer for any AbsGANESPnetModel."""
```
Comment (Masao-Someki) [inline espnet3/systems/codec/gan_trainer.py]: Since this is a public function, let's add more docstrings like Args, Returns, Example!

File: test/espnet3/systems/codec/conftest.py [hunk 12]
```diff
@@ -0,0 +1,14 @@
+import pytest
+
+from espnet3.components.data import data_organizer as data_organizer_module
+
+from ._gan_dummies import DummyDataset
```
Comment (Masao-Someki) [inline test/espnet3/systems/codec/conftest.py]: Let's import with full module path, so that it would be `from test.espnet3.systems.codec._gan_dummies import ..`

File: test/espnet3/systems/codec/test_gan_lightning_module.py [hunk 13]
```diff
@@ -0,0 +1,393 @@
+"""Unit tests for espnet3.systems.codec.models.gan_lightning_module."""
+
+import pytest
+import torch
+import torch.nn as nn
+from omegaconf import OmegaConf
+
+from espnet3.systems.codec.models.gan_lightning_module import GANLightningModule
+
+from ._gan_dummies import (
```
Comment (Masao-Someki) [inline test/espnet3/systems/codec/test_gan_lightning_module.py]: ditto

File: test/espnet3/systems/codec/test_system.py [hunk 14]
```diff
@@ -0,0 +1,209 @@
+"""Unit tests for espnet3.systems.codec.system."""
+
+import pytest
+import torch.nn as nn
+from omegaconf import OmegaConf
+
+import espnet3.systems.codec.system as sysmod
+from espnet3.systems.codec.system import CodecSystem, _instantiate_model
+
+from ._gan_dummies import DummyGANModel
```
Comment (Masao-Someki) [inline test/espnet3/systems/codec/test_system.py]: ditto

**Review by Masao-Someki — COMMENTED** [review]
Thank you @NewGamezzz!!
I've added several comments!

File: egs3/libritts/codec/conf/metrics.yaml (hunk 2 above)
Comment (NewGamezzz) [reply egs3/libritts/codec/conf/metrics.yaml]: Thanks. I agree. Since VERSA metrics can be used by several models, we should move to espnet3 directory.

File: egs3/libritts/codec/conf/training_encodec.yaml (hunk 3 above)
Comment (NewGamezzz) [reply egs3/libritts/codec/conf/training_encodec.yaml]: Done.

File: egs3/libritts/codec/dataset/builder.py (hunk 4 above)
Comment (NewGamezzz) [reply egs3/libritts/codec/dataset/builder.py:207]: Thanks. I removed the bash script for downloading the libritts dataset and reimplement it in `builder.py`.

File: egs3/libritts/codec/dataset/builder.py (hunk 5 above)
Comment (NewGamezzz) [reply egs3/libritts/codec/dataset/builder.py:287]: Done

File: egs3/libritts/codec/local/download_libritts.sh
Comment (NewGamezzz) [reply egs3/libritts/codec/local/download_libritts.sh:1]: Thanks. I moved it to `builder.py`.

File: espnet3/components/metrics/versa.py
Comment (NewGamezzz) [reply espnet3/components/metrics/versa.py:1]: Done.

File: egs3/TEMPLATE/codec/conf/training.yaml (hunk 6 above)
Comment (NewGamezzz) [reply egs3/TEMPLATE/codec/conf/training.yaml]: Sorry about this. I forgot that this is template config. I removed the specific values and move them to the actual recipe's config.

File: egs3/TEMPLATE/codec/conf/training.yaml (hunk 7 above)
Comment (NewGamezzz) [reply egs3/TEMPLATE/codec/conf/training.yaml]: done.

File: egs3/TEMPLATE/codec/conf/training.yaml (hunk 8 above)
Comment (NewGamezzz) [reply egs3/TEMPLATE/codec/conf/training.yaml]: done.

File: egs3/TEMPLATE/codec/conf/training.yaml (hunk 9 above)
Comment (NewGamezzz) [reply egs3/TEMPLATE/codec/conf/training.yaml]: Done.

File: egs3/TEMPLATE/codec/run.py (hunk 10 above)
Comment (NewGamezzz) [reply egs3/TEMPLATE/codec/run.py:30]: Thanks. I have added it.

File: espnet3/systems/codec/gan_trainer.py (hunk 11 above)
Comment (NewGamezzz) [reply espnet3/systems/codec/gan_trainer.py]: Thanks. I makes sure that every public functions follow this format.

File: test/espnet3/systems/codec/conftest.py (hunk 12 above)
Comment (NewGamezzz) [reply test/espnet3/systems/codec/conftest.py]: Done.

File: test/espnet3/systems/codec/test_gan_lightning_module.py (hunk 13 above)
Comment (NewGamezzz) [reply test/espnet3/systems/codec/test_gan_lightning_module.py]: Done.

File: test/espnet3/systems/codec/test_system.py (hunk 14 above)
Comment (NewGamezzz) [reply test/espnet3/systems/codec/test_system.py]: Done.

**Comment by NewGamezzz** [comment]
@Masao-Someki I found a bug when trying to rerun the training stage. The issue was caused by the `EpochSyncIterator` that I introduced previously. I am so sorry about this. I have created a PR to fix the issue. Could you please take a look at this PR first https://github.com/espnet/espnet/pull/6605

**Comment by NewGamezzz** [comment]
@Masao-Someki  I have tested the recipe end to end, and everything is working. I think it is ready to be merged now.

I also added a README, including the publication and demo links. 
- publication: https://huggingface.co/NewGame/libritts_codec_train_encodec_libritts
- demo: https://huggingface.co/spaces/NewGame/libritts_codec_train_encodec_libritts

**Comment by Masao-Someki** [comment]
Hi, based on our discussion on system naming rules, could you change the system name to `esp2_codec`?

**Comment by NewGamezzz** [comment]
Got it. I have changed the directory name to `esp2_codec`. I also changed the README filename to `README.md` to follow your convention.

File: egs3/libritts/esp2_codec/conf/publication.yaml [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+# settings. `exp_tag`/`exp_dir` are copied in from the training config by
+# run.py, so only recipe-specific keys live here.
+
+pack_model:
+  # The template excludes every checkpoint from the bulk copy, so the one that
+  # `conf/inference.yaml` references must be registered here; `files` entries
+  # bypass `exclude` and the packed inference config is rewritten to the bundled
+  # copy. Keep this in sync with `model.model_file` in conf/inference.yaml.
+  files:
+    model_file: ${exp_dir}/epoch95_step1201215_step.ckpt
```
Comment (Masao-Someki) [inline egs3/libritts/esp2_codec/conf/publication.yaml]: Let's avoid using specific model path
